### PR TITLE
get_tld(), get_components(), and other

### PR DESCRIPTION
**Pull request by vadym-t**
Reworked the code of get_tld() to reduce the number of temporary objects (reduce the load on GC in the case of bulk operations). The requirement for the input domain name is updated - it allows to have consistent behavior of all methods. The updated requirement: input domain string is a FQDN without trailing dot. Empty labels are not allowed.

1. get_tld() is split into get_tld() and get_tld_unsafe(). The latter implements the entire logic and the former performs lower() conversion and None checks.
2. get_tld_unsafe() rewritten to create minimal number of temporary objects
3. get_sld() split info get_sld() and get_sld_unsafe(). The latter is based on get_tld_unsafe(), same principles.
4. get_components() and get_components_unsafe() method introduced. These methods allow split the domain string into 3-tuple: (prefix, SLL, TLD) where SLL is second-level label and TLD is TLD or ETLD

### CODE REVIEW

**Review by pombredanne — CHANGES_REQUESTED** [review]
@vadym-t Thank you ++
That's a big PR and I reckon this would change the API and behavior.
I would rather avoid doing so as it would break existing code. Would there be a way to achieve something similar without changing behavior? (e.g. adding new functions or optional args?)

Also you have introduced Python 3-only idioms that make the test fails. For now, I would like to keep things working on Python 2.7 too. Do you think this could be reworked?

**Comment by pombredanne** [comment]
@hiratara @KnitCode ping... your review is much welcomed.

**Comment by KnitCode** [comment]
> @vadym-t Thank you ++
> That's a big PR and I reckon this would change the API and behavior.
> I would rather avoid doing so as it would break existing code. Would there be a way to achieve something similar without changing behavior? (e.g. adding new functions or optional args?)
> 
> Also you have introduced Python 3-only idioms that make the test fails. For now, I would like to keep things working on Python 2.7 too. Do you think this could be reworked?

Hi. I've worked with Vadym extensively over the last week+ on this logic. He is on our team and identified logical bugs and performance issues. I haven't yet code reviewed the latest version, but tested earlier version extensively w/o issue. To your point, though, didn't test python2.7 at all. 

The primary changes are two-fold:
* changes to the computation that reduces the creation of temporary objects. this has the same result as the previous code. In my testing, it was slightly faster than the old version on datasets that would fit in memory for get_tld() and get_sld(). for large distributed sets, the timing was the same but less objects created. this helps for Dask/Spark usage of the library. 
* test changes. vadym dug into the algorithm on the mozilla site and compared to the tests (theirs and ours). there were tests in their suite that are inconsistent and appear to be based on an old PSL list, but passed for us because of default logic in our existing code, which itself was faulty on edge cases. a number of the issues occur at the edge that most users, e.g, web page usage, wouldn't see but we see in very large DNS systems. 

Vadym asked what to do -- and I suggested fix the tests to what they really should be and do a PR. In addition, he added a new suite which starts from the published algorithm and addresses all the combinations. 

So overall, I think it should be merged and will be an upgrade for everyone. the functionality breaks will occur at the edge for very large-scale DNS type things where you'd see bad behavior and root queries.  We plan to adopt in Infoblox Data Science. 

will re-review the code and documentation

File: UseCases.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
+This file describes exact behavior of methods for different edge cases and 
+explains general logic. This description covers the behavior of get_tld, 
+get_tld_unsafe, get_sld, get_sld_unsafe, split_domain, split_domain_unsafe
+
```
Comment (KnitCode) [inline UseCases.md:4]: @vadym-t suggest to add here a why sentence: 
Unsafe versions of the methods will significantly save resources on large-scale applications of the library where the data has already been converted to lowercase and missing data has a None value. This can be done in Spark/Dask, for example, and result in a significant reduction in computational resources. For adhoc usage, the original functions are sufficient.

File: UseCases.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+4. Every method processes provided FQDN in the reverse order, from the last 
+label towards the start of the string. It stops when the specific task is 
+completed. Therefore no validation occurs outside of this scope.  For example,
+```
+get_tld('......com') -> 'com'
+``` 
+as leading dots are not processed.
+split_domain method is based on get_sld method - it returns everything in 
+front of get_sld() as a prefix.
+Specifically to example above
```
Comment (KnitCode) [inline UseCases.md:31]: @vadym-t you might also add here a non-edge case example. the split_domain() method offers a new capability to the library-- one that folks might get from other libraries -- but your only example is the edge. suggest something like:
split_domain allows you to recover the host, or prefix, of an SLD, for use in aggregation or analysis based on the labels. e.g., split_domain('www.googl.com')

File: UseCases.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
+This file describes exact behavior of methods for different edge cases and 
+explains general logic. This description covers the behavior of get_tld, 
+get_tld_unsafe, get_sld, get_sld_unsafe, split_domain, split_domain_unsafe
+
+1. general difference of get_*() and get_*_unsafe() methods:
+get_*_unsafe() does not perform if the input string is None and does not 
+transforms it to the lower case.
+
+2. The listed above methods works only with non-canonical FQDN strings -
```
Comment (KnitCode) [inline UseCases.md:15]: listed above means all or just the unsafe methods?

File: UseCases.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+split_domain method is based on get_sld method - it returns everything in 
+front of get_sld() as a prefix.
+Specifically to example above 
+```
+split_domain('......com') -> ('....',None,'com')
+```
+Edge cases and expected behavior
+The behavior of the library can be illustrated best on the small examples:
+(boolean arguments are omitted if does not affect behavior )
+
```
Comment (KnitCode) [inline UseCases.md:38]: Suggested additional: 
The library allows you to create a public suffix list from any file-like object, including a list. In the examples below, we construct the PSL with different lists to demonstrate the functionality of the library under different conditions.  In particular, the examples show an empty list, list of two elements, and the use of negation in a list. Combinations of the method parameters are given with notes about the result.  These results follow the logic of the Mozilla library as published on the psl.org website.

File: tests.py [hunk 5]
```diff
@@ -78,7 +81,8 @@ def test_get_sld_from_list_with_exception_rule(self):
 
     def test_get_sld_from_list_with_fqdn(self):
         psl = publicsuffix.PublicSuffixList(['com'])
-        assert 'example.com' == psl.get_sld('example.com.')
+        # 'example.com.' -> <example>.<com>.<empty> -> None, empty labels are not allowed
+        assert None == psl.get_sld('example.com.')
```
Comment (KnitCode) [inline tests.py:85]: didn't follow this comment. is this failing because of the trailing dot? I read the code as the get_sld() version will remove the trailing dot ... so why doesn't this go to .. 
example.com -> example, com -> example.com 
??

File: tests.py [hunk 6]
```diff
@@ -49,8 +49,11 @@ def test_get_sld_from_empty_list(self):
         psl = publicsuffix.PublicSuffixList([])
         assert 'com' == psl.get_sld('com')
         assert 'com' == psl.get_sld('COM')
-        assert 'com' == psl.get_sld('.com')
-        assert 'com' == psl.get_sld('a.example.com')
+        # '.com' -> <empty>.<com> -> None, empty labels are not allowed
+        assert None == psl.get_sld('.com')
```
Comment (KnitCode) [inline tests.py:53]: since this is the safe, not unsafe version, my expectation is that the dots are stripped. is this not the case? so '.com' becomes 'com' before being processed. it doesn't have a second label, but it's not empty either. previously we'd return 'com' for this. 

the sld from empty list is essentially returning two labels. there are use cases for this, even though it seems like a bizarre thing to do. so the example.com example is good here too.

File: tests.py [hunk 7]
```diff
@@ -103,24 +107,29 @@ def test_get_sld_from_builtin_full_publicsuffix_org_list_with_mixed_case(self):
 
     def test_get_sld_from_builtin_full_publicsuffix_org_list_with_leading_dot(self):
         psl = publicsuffix.PublicSuffixList(None)
-        assert 'com' == psl.get_sld('.com')
-        assert 'example' == psl.get_sld('.example')
+        # '.com' -> <empty>.<com> -> None, empty labels are not allowed
```
Comment (KnitCode) [inline tests.py:110]: same confusion. if get_sld() does strip '.' on both sides, why are these test results different?

File: tests.py [hunk 8]
```diff
@@ -103,24 +107,29 @@ def test_get_sld_from_builtin_full_publicsuffix_org_list_with_mixed_case(self):
...
+        assert 'example.example' == psl.get_sld('b.example.example')
+        assert 'example.example' == psl.get_sld('a.b.example.example')
 
     def test_get_sld_from_builtin_full_publicsuffix_org_list_with_listed_ut_non_internet_tld(self):
         psl = publicsuffix.PublicSuffixList(None)
         assert 'local' == psl.get_sld('local')
-        assert 'local' == psl.get_sld('example.local')
-        assert 'local' == psl.get_sld('b.example.local')
-        assert 'local' == psl.get_sld('a.b.example.local')
+        # note: non-strict mode, unknown tld: local -> sld: example.local
```
Comment (KnitCode) [inline tests.py:129]: my thoughts are that this whole suite might be removed. local is not listed in the psl anymore -- though i believe it was once and that's how the test was made. this series of tests assumes something is true that isn't in our control -- the psl.  i'm not sure what this unit does at all to help.

File: tests.py [hunk 9]
```diff
@@ -237,16 +246,19 @@ def test_get_tld_returns_correct_tld_or_etld(self):
...
         assert None == psl.get_tld('blah.local', strict=True)
-        assert None == psl.get_tld('blah.local', wildcard=False)
+        # non-strict mode: TLD = 'local'
+        assert 'local' == psl.get_tld('blah.local', wildcard=False)
         assert 'local' == psl.get_tld('blah.local')
 
     def test_get_tld_returns_correct_tld_or_etld_for_fqdn(self):
         psl = publicsuffix.PublicSuffixList()
-        assert 'com' == psl.get_tld('www.foo.com.')
+        # note: empty label or dot on the right side is not allowed
```
Comment (KnitCode) [inline tests.py:255]: i thought that was just for the unsafe version of the method? get_tld() would strip but unsafe would not.

File: tests.py [hunk 10]
```diff
@@ -270,11 +282,12 @@ def test_get_sld_backward_compatibility(self):
     def test_get_sld_backward_compatibility_strict_and_wildcard_flags(self):
         psl = publicsuffix.PublicSuffixList()
         assert 'local' == psl.get_sld('local')
-        assert 'local' == psl.get_sld('foo.local')
+        # non-strict mode: TLD=local -> sld = foo.local
+        assert 'foo.local' == psl.get_sld('foo.local')
```
Comment (KnitCode) [inline tests.py:286]: this is logically correct, though not backward compatible. this is one of the old tests that didn't match the logic of alg.

File: tests.py [hunk 11]
```diff
@@ -284,13 +297,15 @@ def test_get_sld_backward_compatibility_sld_for_empty_string(self):
 
     def test_get_sld_backward_compatibility_sld_for_fqdn(self):
         psl = publicsuffix.PublicSuffixList()
-        assert 'foo.com' == psl.get_sld('www.foo.com.')
+        # 'www.foo.com.' -> <www>.<foo>.<com>.<empty> -> None, empty labels are not allowed
+        assert None == psl.get_sld('www.foo.com.')
```
Comment (KnitCode) [inline tests.py:301]: same question abut strip

i don't see any tests for the unsafe methods. shouldn't there be a few? this is what the big change is.

File: tests.py [hunk 12]
```diff
@@ -284,13 +297,15 @@ def test_get_sld_backward_compatibility_sld_for_empty_string(self):
...
 
     def test_get_sld_backward_compatibility_sld_for_root_domain(self):
         psl = publicsuffix.PublicSuffixList()
-        assert '' == psl.get_sld('.')
+        # empty labels are not allowed
+        assert None == psl.get_sld('.')
         assert None == psl.get_sld('.', strict=True)
-        assert '' == psl.get_sld('.', wildcard=False)
+        assert None == psl.get_sld('.', wildcard=False)
```
Comment (KnitCode) [inline tests.py:308]: not backward compatible but logically more accurate

File: tests_mozilla.py [hunk 13]
```diff
@@ -583,40 +583,49 @@ def test_get_sld_Mixed_case3(self):
         assert 'example.com' == publicsuffix.get_sld('WwW.example.COM')
 
     def test_get_sld_Leading_dot1(self):
-        assert 'com' == publicsuffix.get_sld('.com')
+        # '.com' -> <empty>.<com> -> None, empty labels are not allowed
+        assert None == publicsuffix.get_sld('.com')
```
Comment (KnitCode) [inline tests_mozilla.py:587]: i'm going to continue having all the strip questions - so i won't repeat them.

File: tests_mozilla.py [hunk 14]
```diff
@@ -583,40 +583,49 @@ def test_get_sld_Mixed_case3(self):
...
 
     def test_get_sld_Leading_dot2(self):
-        assert 'example' == publicsuffix.get_sld('.example')
+        # '.example' -> <empty>.<example> -> None, empty labels are not allowed
+        assert None == publicsuffix.get_sld('.example')
```
Comment (KnitCode) [inline tests_mozilla.py:591]: the difference here is that example is not in the list. so in strict mode should return None and in non-strict mode would return example --- with the strip

correct?

File: tests_mozilla.py [hunk 15]
```diff
@@ -583,40 +583,49 @@ def test_get_sld_Mixed_case3(self):
...
+        # non-strict: TLD:example, SLD:example.example
+        assert 'example.example' == publicsuffix.get_sld('.example.example')
 
     def test_get_sld_Unlisted_sld1(self):
         assert 'example' == publicsuffix.get_sld('example')
 
     def test_get_sld_Unlisted_sld2(self):
-        assert 'example' == publicsuffix.get_sld('example.example')
+        # non-strict: TLD:example, SLD:example.example
+        assert 'example.example' == publicsuffix.get_sld('example.example')
```
Comment (KnitCode) [inline tests_mozilla.py:605]: this is a change where the original logic was flawed -- didn't follow the algorithm complete-- so the test itself was incorrect

**Review by KnitCode — CHANGES_REQUESTED** [review]
my main comments are on:
* documentation suggestions to add a few things, 
* in the tests, where i'm expecting a strip to occur in the function get_sld(), it seems the tests do not. also, i don't see any tests for the unsafe versions --that's where i'd expect to see the none behavior. 

am i missing something? 

where the backward compatibility breaks are
* edge cases that are rare, e.g., having root return None instead of empty string -- don't have strong opinion on this. 
* cases where the tests didn't match the published algorithm

**Comment by hiratara** [comment]
`_lookup_node` doesn't seem to be used anymore.

https://github.com/nexB/python-publicsuffix2/blob/e95f43d26d57692fc1f153c8c9bbde62dabeaae7/src/publicsuffix2/__init__.py#L201-L240

I'm all for improving the execution speed. However, we are cautious about changing existing behavior. The reason the specs of this library change so often is that the specs are so vague. It's hard to determine if this specification change is reasonable (although I'm not actively opposed to it).

File: UseCases.md (hunk 1 above)
Comment (vadym-t) [reply UseCases.md:4]: Added

**Comment by KnitCode** [comment]
> `_lookup_node` doesn't seem to be used anymore.
> 
> https://github.com/nexB/python-publicsuffix2/blob/e95f43d26d57692fc1f153c8c9bbde62dabeaae7/src/publicsuffix2/__init__.py#L201-L240
> 
> I'm all for improving the execution speed. However, we are cautious about changing existing behavior. The reason the specs of this library change so often is that the specs are so vague. It's hard to determine if this specification change is reasonable (although I'm not actively opposed to it).

agree, and my brain hadn't noticed that. @pombredanne  I would think the priorities would be in rough order: 
* correcting faulty logic and unit tests based on the published algorithm; these corrections do a good job of documenting the algorithm and reveal faults in the original Mozilla tests that rely on the existence of certain things (e.g. example) being the psl which aren't.  i like the demonstrated combinations using a set list in the use_caeses.md 
* ensuring backward compatibility, outside of being wrong :) ...  so that the functions people are using today work as expected, and added functions give more controls, as needed, 
* py2.7 support for older use cases, 
* performance improvements.. that's always a fine thing. 

agreed? given that, wondering if the merge can be split to get these.  if the performance improvements suggested by @vadym-t only work with py3 idioms and don't use the trie, what about using an extra flag parameter to trigger usage of the new functions? 

i can take a closer look on the weekend and rerun the experiments i did with an earlier version.

**Comment by pombredanne** [comment]
@vadym-t ping?
